File: backend/app/core/laporan_penerimaan.py
"""
Logic Laporan Penerimaan. Berbeda dari app/core/laporan.py (yang query
jurnal_detail) -- laporan di sini query langsung dari tabel Tagihan dan
PembayaranTagihan, karena yang dibutuhkan adalah sudut pandang siswa/kelas,
bukan akun jurnal. Tetap live query, tidak ada tabel hasil hitung terpisah.

Definisi "siswa menunggak" untuk MVP: Tagihan.status == BELUM_LUNAS dan
sisa > 0. Model Tagihan belum punya field jatuh tempo, jadi belum bisa
membedakan "menunggak" vs "belum jatuh tempo" -- cukup untuk MVP.
"""
from decimal import Decimal
from typing import Optional

from sqlalchemy import func
from sqlalchemy.orm import Session

from app.models.penerimaan import PembayaranTagihan, StatusPembayaran, StatusTagihan, Tagihan
from app.models.siswa import Kelas, Siswa


def tagihan_per_kelas(db: Session, tahun_buku_id: int, departemen_id: Optional[int] = None) -> list[dict]:
    """Rekap tagihan per kelas: total ditagihkan, diskon, sisa, jumlah tagihan."""
    query = (
        db.query(
            Kelas.id, Kelas.nama, Kelas.tingkat,
            func.coalesce(func.sum(Tagihan.jumlah_tagihan), 0),
            func.coalesce(func.sum(Tagihan.diskon), 0),
            func.coalesce(func.sum(Tagihan.sisa), 0),
            func.count(Tagihan.id),
        )
        .join(Siswa, Siswa.id == Tagihan.siswa_id)
        .join(Kelas, Kelas.id == Siswa.kelas_id)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(Tagihan.status != StatusTagihan.DIBATALKAN)
    )
    if departemen_id is not None:
        query = query.filter(Tagihan.departemen_id == departemen_id)
    rows = query.group_by(Kelas.id, Kelas.nama, Kelas.tingkat).order_by(Kelas.tingkat, Kelas.nama).all()
    return [
        {
            "kelas_id": kelas_id, "kelas_nama": nama, "tingkat": tingkat,
            "total_tagihan": total_tagihan, "total_diskon": total_diskon,
            "total_sisa": total_sisa, "jumlah_tagihan": jumlah_tagihan,
        }
        for kelas_id, nama, tingkat, total_tagihan, total_diskon, total_sisa, jumlah_tagihan in rows
    ]


def tagihan_per_siswa_detail(db: Session, tahun_buku_id: int, siswa_id: int) -> list[dict]:
    """Detail semua tagihan satu siswa di tahun buku tertentu."""
    rows = (
        db.query(Tagihan)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(Tagihan.siswa_id == siswa_id)
        .order_by(Tagihan.tanggal)
        .all()
    )
    return [
        {
            "tagihan_id": t.id, "no_tagihan": t.no_tagihan, "tanggal": t.tanggal,
            "jenis_pembayaran_nama": t.jenis_pembayaran.nama,
            "jumlah_tagihan": t.jumlah_tagihan, "diskon": t.diskon,
            "sisa": t.sisa, "status": t.status,
        }
        for t in rows
    ]


def tagihan_rekap_per_siswa(db: Session, tahun_buku_id: int, departemen_id: Optional[int] = None) -> list[dict]:
    """Rekap total tagihan per siswa (tanpa drill-down ke tagihan_id)."""
    query = (
        db.query(
            Siswa.id, Siswa.nis, Siswa.nama, Kelas.nama,
            func.coalesce(func.sum(Tagihan.jumlah_tagihan), 0),
            func.coalesce(func.sum(Tagihan.diskon), 0),
            func.coalesce(func.sum(Tagihan.sisa), 0),
        )
        .join(Tagihan, Tagihan.siswa_id == Siswa.id)
        .join(Kelas, Kelas.id == Siswa.kelas_id)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(Tagihan.status != StatusTagihan.DIBATALKAN)
    )
    if departemen_id is not None:
        query = query.filter(Tagihan.departemen_id == departemen_id)
    rows = query.group_by(Siswa.id, Siswa.nis, Siswa.nama, Kelas.nama).order_by(Siswa.nama).all()
    return [
        {
            "siswa_id": siswa_id, "nis": nis, "nama": nama, "kelas_nama": kelas_nama,
            "total_tagihan": total_tagihan, "total_diskon": total_diskon, "total_sisa": total_sisa,
        }
        for siswa_id, nis, nama, kelas_nama, total_tagihan, total_diskon, total_sisa in rows
    ]


def _query_siswa_menunggak(db: Session, tahun_buku_id: int, departemen_id: Optional[int], kelas_id: Optional[int]):
    query = (
        db.query(Tagihan)
        .join(Siswa, Siswa.id == Tagihan.siswa_id)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(Tagihan.status == StatusTagihan.BELUM_LUNAS)
        .filter(Tagihan.sisa > 0)
    )
    if departemen_id is not None:
        query = query.filter(Tagihan.departemen_id == departemen_id)
    if kelas_id is not None:
        query = query.filter(Siswa.kelas_id == kelas_id)
    return query


def siswa_menunggak_detail(
    db: Session, tahun_buku_id: int, departemen_id: Optional[int] = None, kelas_id: Optional[int] = None
) -> list[dict]:
    """Daftar tagihan yang masih menunggak (BELUM_LUNAS, sisa > 0), satu baris per tagihan."""
    rows = _query_siswa_menunggak(db, tahun_buku_id, departemen_id, kelas_id).order_by(Tagihan.tanggal).all()
    return [
        {
            "tagihan_id": t.id, "no_tagihan": t.no_tagihan, "tanggal": t.tanggal,
            "siswa_id": t.siswa_id, "siswa_nama": t.siswa.nama, "siswa_nis": t.siswa.nis,
            "kelas_nama": t.siswa.kelas.nama,
            "jenis_pembayaran_nama": t.jenis_pembayaran.nama,
            "sisa": t.sisa,
        }
        for t in rows
    ]


def siswa_menunggak_rekap(
    db: Session, tahun_buku_id: int, departemen_id: Optional[int] = None
) -> list[dict]:
    """Rekap total tunggakan per siswa (dipakai dashboard & rekap laporan)."""
    query = (
        db.query(
            Siswa.id, Siswa.nis, Siswa.nama, Kelas.nama,
            func.coalesce(func.sum(Tagihan.sisa), 0),
            func.count(Tagihan.id),
        )
        .join(Tagihan, Tagihan.siswa_id == Siswa.id)
        .join(Kelas, Kelas.id == Siswa.kelas_id)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(Tagihan.status == StatusTagihan.BELUM_LUNAS)
        .filter(Tagihan.sisa > 0)
    )
    if departemen_id is not None:
        query = query.filter(Tagihan.departemen_id == departemen_id)
    rows = query.group_by(Siswa.id, Siswa.nis, Siswa.nama, Kelas.nama).order_by(Siswa.nama).all()
    return [
        {
            "siswa_id": siswa_id, "nis": nis, "nama": nama, "kelas_nama": kelas_nama,
            "total_sisa": total_sisa, "jumlah_tagihan_menunggak": jumlah_tagihan,
        }
        for siswa_id, nis, nama, kelas_nama, total_sisa, jumlah_tagihan in rows
    ]


def tunggakan_per_kelas(db: Session, tahun_buku_id: int, departemen_id: Optional[int] = None) -> list[dict]:
    """Rekap total tunggakan per kelas -- dipakai bar chart Dashboard."""
    query = (
        db.query(
            Kelas.id, Kelas.nama,
            func.coalesce(func.sum(Tagihan.sisa), 0),
        )
        .join(Siswa, Siswa.id == Tagihan.siswa_id)
        .join(Kelas, Kelas.id == Siswa.kelas_id)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(Tagihan.status == StatusTagihan.BELUM_LUNAS)
        .filter(Tagihan.sisa > 0)
    )
    if departemen_id is not None:
        query = query.filter(Tagihan.departemen_id == departemen_id)
    rows = query.group_by(Kelas.id, Kelas.nama).order_by(Kelas.nama).all()
    return [
        {"kelas_id": kelas_id, "kelas_nama": nama, "total_sisa": total_sisa}
        for kelas_id, nama, total_sisa in rows
    ]


def rekapitulasi_penerimaan(
    db: Session,
    tahun_buku_id: int,
    departemen_id: Optional[int] = None,
    tanggal_mulai=None,
    tanggal_selesai=None,
) -> list[dict]:
    """Rekap per jenis pembayaran: total ditagihkan (Tagihan) vs total
    diterima tunai (PembayaranTagihan status AKTIF), opsional difilter rentang
    tanggal pembayaran."""
    from app.models.penerimaan import JenisPembayaran

    tagihan_query = (
        db.query(
            JenisPembayaran.id, JenisPembayaran.nama,
            func.coalesce(func.sum(Tagihan.jumlah_tagihan - Tagihan.diskon), 0),
        )
        .join(Tagihan, Tagihan.jenis_pembayaran_id == JenisPembayaran.id)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(Tagihan.status != StatusTagihan.DIBATALKAN)
    )
    if departemen_id is not None:
        tagihan_query = tagihan_query.filter(Tagihan.departemen_id == departemen_id)
    tagihan_rows = tagihan_query.group_by(JenisPembayaran.id, JenisPembayaran.nama).all()
    ditagihkan_map = {jp_id: (nama, total) for jp_id, nama, total in tagihan_rows}

    diterima_query = (
        db.query(
            JenisPembayaran.id,
            func.coalesce(func.sum(PembayaranTagihan.jumlah_bayar), 0),
        )
        .join(Tagihan, Tagihan.jenis_pembayaran_id == JenisPembayaran.id)
        .join(PembayaranTagihan, PembayaranTagihan.tagihan_id == Tagihan.id)
        .filter(Tagihan.tahun_buku_id == tahun_buku_id)
        .filter(PembayaranTagihan.status == StatusPembayaran.AKTIF)
    )
    if departemen_id is not None:
        diterima_query = diterima_query.filter(Tagihan.departemen_id == departemen_id)
    if tanggal_mulai is not None:
        diterima_query = diterima_query.filter(PembayaranTagihan.tanggal >= tanggal_mulai)
    if tanggal_selesai is not None:
        diterima_query = diterima_query.filter(PembayaranTagihan.tanggal <= tanggal_selesai)
    diterima_rows = diterima_query.group_by(JenisPembayaran.id).all()
    diterima_map = {jp_id: total for jp_id, total in diterima_rows}

    semua_id = set(ditagihkan_map) | set(diterima_map)
    hasil = []
    for jp_id in semua_id:
        nama, total_ditagihkan = ditagihkan_map.get(jp_id, (None, Decimal("0")))
        if nama is None:
            from app.models.penerimaan import JenisPembayaran as JP
            nama = db.query(JP.nama).filter(JP.id == jp_id).scalar()
        total_diterima = diterima_map.get(jp_id, Decimal("0"))
        hasil.append({
            "jenis_pembayaran_id": jp_id, "jenis_pembayaran_nama": nama,
            "total_ditagihkan": total_ditagihkan, "total_diterima": total_diterima,
        })
    hasil.sort(key=lambda r: r["jenis_pembayaran_nama"] or "")
    return hasil

File: frontend/static/js/modules/laporan-penerimaan.js
/**
 * Modul Laporan Penerimaan — laporan ~10 jenis dari sudut pandang siswa/kelas
 * (beda dari LaporanModule yang query jurnal_detail). Mengikuti pola tab
 * persis seperti ReferensiModule/LaporanModule.
 *
 * Tab siswa-menunggak baca params[1] sebagai kelas_id opsional (dipakai saat
 * navigasi dari Dashboard lewat klik bar chart: hash
 * #laporan-penerimaan/siswa-menunggak/<kelas_id> -> params[0]='siswa-menunggak',
 * params[1]='<kelas_id>').
 */
class LaporanPenerimaanModule {
  constructor(container, api, params) {
    this.container = container;
    this.api = api;
    this.tab = params[0] || 'per-kelas'; // per-kelas | per-siswa | siswa-menunggak | rekapitulasi
    this._kelasIdAwal = this.tab === 'siswa-menunggak' && params[1] ? Number(params[1]) : null;
    this._departemenCache = null;
    this._tahunBukuCache = null;
    this._siswaCache = null;
    this._kelasCache = null;
  }

  async render() {
    this.container.innerHTML = `
      <div class="page-header">
        <div class="breadcrumb">Laporan Penerimaan</div>
        <div class="page-title"><h1>${this._judulTab()}</h1></div>
      </div>
      <div class="tabs" style="display:flex; gap: var(--space-2); margin-bottom: var(--space-4); flex-wrap: wrap;">
        ${this._renderTabButton('per-kelas', 'Per Kelas')}
        ${this._renderTabButton('per-siswa', 'Per Siswa')}
        ${this._renderTabButton('siswa-menunggak', 'Siswa Menunggak')}
        ${this._renderTabButton('rekapitulasi', 'Rekapitulasi')}
      </div>
      <div class="panel">
        <div class="panel__header">
          <div class="flex-between" style="gap: var(--space-3); flex: 1;">
            <span class="text-muted">Daftar ${this._judulTab()}</span>
            <div id="filter-container" style="display:flex; gap: var(--space-2);"></div>
          </div>
          <button class="btn btn-outline btn-sm no-print" id="btn-cetak">🖨 Cetak / Export PDF</button>
        </div>
        <div class="panel__body" id="tab-content">
          <div class="empty-state">Memuat data...</div>
        </div>
      </div>
    `;

    this.container.querySelectorAll('[data-tab]').forEach(btn => {
      btn.addEventListener('click', () => router.navigate(`laporan-penerimaan/${btn.dataset.tab}`));
    });
    this.container.querySelector('#btn-cetak').addEventListener('click', () => this._cetakLaporanAktif());

    await this._renderFilter();
    await this._muatData();
  }

  _judulTab() {
    return {
      'per-kelas': 'Tagihan per Kelas',
      'per-siswa': 'Tagihan per Siswa',
      'siswa-menunggak': 'Siswa Menunggak',
      rekapitulasi: 'Rekapitulasi Penerimaan',
    }[this.tab];
  }

  _renderTabButton(key, label) {
    const active = this.tab === key ? 'btn-primary' : 'btn-outline';
    return `<button class="btn ${active} btn-sm" data-tab="${key}">${label}</button>`;
  }

  async _muatDepartemen() {
    if (!this._departemenCache) this._departemenCache = await this.api.get('/referensi/departemen');
    return this._departemenCache;
  }

  async _muatTahunBuku() {
    if (!this._tahunBukuCache) this._tahunBukuCache = await this.api.get('/referensi/tahun-buku');
    return this._tahunBukuCache;
  }

  async _muatSiswa() {
    if (!this._siswaCache) this._siswaCache = await this.api.get('/siswa/siswa');
    return this._siswaCache;
  }

  async _muatKelas() {
    if (!this._kelasCache) this._kelasCache = await this.api.get('/siswa/kelas');
    return this._kelasCache;
  }

  _tahunBukuAktif(daftar) {
    return daftar.find(t => t.status === 'AKTIF');
  }

  async _renderFilter() {
    const filterEl = this.container.querySelector('#filter-container');
    const tahunBuku = await this._muatTahunBuku();
    const aktif = this._tahunBukuAktif(tahunBuku);
    if (!aktif) {
      filterEl.innerHTML = '';
      this._tahunBukuId = null;
      return;
    }
    this._tahunBukuId = aktif.id;

    let extraHtml = '';
    if (this.tab === 'per-siswa') {
      const siswa = await this._muatSiswa();
      extraHtml = `
        <select id="filter-siswa">
          <option value="">-- pilih siswa untuk detail --</option>
          ${siswa.map(s => `<option value="${s.id}">${s.nis} - ${s.nama}</option>`).join('')}
        </select>
      `;
    } else if (this.tab === 'siswa-menunggak') {
      const kelas = await this._muatKelas();
      extraHtml = `
        <select id="filter-kelas">
          <option value="">-- semua kelas --</option>
          ${kelas.map(k => `<option value="${k.id}" ${this._kelasIdAwal === k.id ? 'selected' : ''}>${k.nama}</option>`).join('')}
        </select>
      `;
    }

    filterEl.innerHTML = `<span class="text-muted">Tahun Buku: ${aktif.tahun_buku}</span>${extraHtml}`;

    const filterSiswa = filterEl.querySelector('#filter-siswa');
    if (filterSiswa) filterSiswa.addEventListener('change', () => this._muatData());
    const filterKelas = filterEl.querySelector('#filter-kelas');
    if (filterKelas) filterKelas.addEventListener('change', () => this._muatData());
  }

  async _muatData() {
    const target = this.container.querySelector('#tab-content');
    if (!this._tahunBukuId) {
      target.innerHTML = '<div class="empty-state">Belum ada Tahun Buku aktif. Buat dulu di modul Referensi.</div>';
      return;
    }
    try {
      if (this.tab === 'per-kelas') {
        const data = await this.api.get(`/laporan-penerimaan/per-kelas?tahun_buku_id=${this._tahunBukuId}`);
        target.innerHTML = TableRenderer.render([
          { key: 'kelas_nama', label: 'Kelas' },
          { key: 'tingkat', label: 'Tingkat' },
          { key: 'total_tagihan', label: 'Total Tagihan', type: 'uang' },
          { key: 'total_diskon', label: 'Total Diskon', type: 'uang' },
          { key: 'total_sisa', label: 'Total Sisa', type: 'uang' },
          { key: 'jumlah_tagihan', label: 'Jumlah Tagihan' },
        ], data, { emptyMessage: 'Belum ada tagihan di tahun buku ini.', searchable: true, searchPlaceholder: 'Cari kelas...' });
        TableRenderer.attachSearch(target);
      } else if (this.tab === 'per-siswa') {
        const filterSiswa = this.container.querySelector('#filter-siswa');
        const siswaId = filterSiswa ? filterSiswa.value : '';
        if (siswaId) {
          const data = await this.api.get(`/laporan-penerimaan/per-siswa/${siswaId}?tahun_buku_id=${this._tahunBukuId}`);
          target.innerHTML = TableRenderer.render([
            { key: 'no_tagihan', label: 'No. Tagihan' },
            { key: 'tanggal', label: 'Tanggal' },
            { key: 'jenis_pembayaran_nama', label: 'Jenis Pembayaran' },
            { key: 'jumlah_tagihan', label: 'Jumlah', type: 'uang' },
            { key: 'diskon', label: 'Diskon', type: 'uang' },
            { key: 'sisa', label: 'Sisa', type: 'uang' },
            { key: 'status', label: 'Status', type: 'badge', badgeMap: {
                BELUM_LUNAS: 'warning', LUNAS: 'success', DIBATALKAN: 'danger',
            } },
          ], data, { emptyMessage: 'Siswa ini belum punya tagihan di tahun buku ini.' });
        } else {
          const data = await this.api.get(`/laporan-penerimaan/per-siswa?tahun_buku_id=${this._tahunBukuId}`);
          target.innerHTML = TableRenderer.render([
            { key: 'nis', label: 'NIS' },
            { key: 'nama', label: 'Nama' },
            { key: 'kelas_nama', label: 'Kelas' },
            { key: 'total_tagihan', label: 'Total Tagihan', type: 'uang' },
            { key: 'total_diskon', label: 'Total Diskon', type: 'uang' },
            { key: 'total_sisa', label: 'Total Sisa', type: 'uang' },
          ], data, { emptyMessage: 'Belum ada tagihan di tahun buku ini.', searchable: true, searchPlaceholder: 'Cari nama/NIS siswa...' });
          TableRenderer.attachSearch(target);
        }
      } else if (this.tab === 'siswa-menunggak') {
        const filterKelas = this.container.querySelector('#filter-kelas');
        const kelasId = filterKelas ? filterKelas.value : '';
        const qs = kelasId ? `&kelas_id=${kelasId}` : '';
        const data = await this.api.get(`/laporan-penerimaan/siswa-menunggak?tahun_buku_id=${this._tahunBukuId}${qs}`);
        target.innerHTML = TableRenderer.render([
          { key: 'no_tagihan', label: 'No. Tagihan' },
          { key: 'tanggal', label: 'Tanggal' },
          { key: 'siswa_nis', label: 'NIS' },
          { key: 'siswa_nama', label: 'Nama Siswa' },
          { key: 'kelas_nama', label: 'Kelas' },
          { key: 'jenis_pembayaran_nama', label: 'Jenis Pembayaran' },
          { key: 'sisa', label: 'Sisa', type: 'uang' },
        ], data, { emptyMessage: 'Tidak ada siswa menunggak.', searchable: true, searchPlaceholder: 'Cari nama/NIS siswa...' });
        TableRenderer.attachSearch(target);
      } else if (this.tab === 'rekapitulasi') {
        const data = await this.api.get(`/laporan-penerimaan/rekapitulasi?tahun_buku_id=${this._tahunBukuId}`);
        target.innerHTML = TableRenderer.render([
          { key: 'jenis_pembayaran_nama', label: 'Jenis Pembayaran' },
          { key: 'total_ditagihkan', label: 'Total Ditagihkan', type: 'uang' },
          { key: 'total_diterima', label: 'Total Diterima', type: 'uang' },
        ], data, { emptyMessage: 'Belum ada data penerimaan di tahun buku ini.' });
      }
    } catch (err) {
      target.innerHTML = `<div class="empty-state">Gagal memuat data: ${err.message}</div>`;
    }
  }

  _cetakLaporanAktif() {
    const tabContent = this.container.querySelector('#tab-content');
    const tanggalCetak = new Date().toLocaleDateString('id-ID');
    const titleHtml = `<h2>${this._judulTab()}</h2><p>Dicetak tanggal: ${tanggalCetak}</p>`;
    PrintHelper.cetak(titleHtml, tabContent.innerHTML);
  }

  destroy() {
    // Tidak ada listener global (interval/websocket) di modul ini, jadi kosong.
  }
}
